### PR TITLE
Refactor TwoDomainLoadBalancer class for separate LoadBalancers

### PR DESCRIPTION
**Pull request by xiancao**
Currently TwoDomainLoadBalancer class has all the LoadBalancer usecase in one class and the Test methods has unnecessary dependencies on each other. In order to avoid dependency, 
Create separate class for each LB. Move the Test(1) and Test(2) to a separate class. For each LoadBalancer class beforeAll() create two domains and install the LoadBalancer, then all tests are independent.

Jenkins job:
https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/8200/
https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/8201/

### CODE REVIEW

**Review by anpanigr — COMMENTED** [review]
Should we rename the test classes to reflect Two Domains Configuration
ItLbTraefik.java --> ItLbTwoDomainsTraefik.java
Same for all new classes.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+    createRouteForOKD("external-weblogic-operator-svc", opNamespace);
+    // Patch the route just created to set tls termination to passthrough
+    setTlsTerminationForRoute("external-weblogic-operator-svc", opNamespace);
+
+    // create pull secrets for WebLogic image when running in non Kind Kubernetes cluster
+    // this secret is used only for non-kind cluster
+    createSecretForBaseImages(domainNamespace);
+
+    for (int i = 1; i <= numberOfDomains; i++) {
+      domainUids.add("wls-domain" + i);
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java:118]: Make the domain id unique across tests modify the prefix "wls-domain"  to "wls-ngix-domain-"

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+
+    // build and deploy app to be used by all test cases
+    buildAndDeployClusterviewApp(domainNamespace, domainUids);
+
+    // install Nginx ingress controller for all test cases using Nginx
+    installIngressController();
+  }
+
+  /**
+   * Verify WebLogic admin console is accessible through NGINX path routing with HTTPS protocol.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java:133]: Modify the description:  Verify the WebLogic Administration Console form both domain is accessible thru a path routing based single NGIX  LoadBalancer  using HTTP protocol.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+import static oracle.weblogic.kubernetes.utils.OKDUtils.setTlsTerminationForRoute;
+import static oracle.weblogic.kubernetes.utils.OperatorUtils.installAndVerifyOperator;
+import static oracle.weblogic.kubernetes.utils.SecretUtils.createSecretWithTLSCertKey;
+import static oracle.weblogic.kubernetes.utils.ThreadSafeLogger.getLogger;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assumptions.assumeFalse;
+
+/**
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java:45]: Modify the description specific to Traefik LoadBalancer.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    createRouteForOKD("external-weblogic-operator-svc", opNamespace);
+    // Patch the route just created to set tls termination to passthrough
+    setTlsTerminationForRoute("external-weblogic-operator-svc", opNamespace);
+
+    // create pull secrets for WebLogic image when running in non Kind Kubernetes cluster
+    // this secret is used only for non-kind cluster
+    createSecretForBaseImages(domainNamespace);
+
+    for (int i = 1; i <= numberOfDomains; i++) {
+      domainUids.add("wls-domain" + i);
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java:112]: Make the domain id unique across lb test classes.

**Comment by maggiehe00** [comment]
Agree with @anpanigr right now we don't run external LB on OKD so no corresponding OKD code are needed in It(LB) test classes.

**Review by maggiehe00 — CHANGES_REQUESTED** [review]
Agree with @anpanigr right now we don't run external LB on OKD so no corresponding OKD code are needed in It(LB) test classes.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+      verifyClusterLoadbalancing(domainUids.get(i - 1), "", "http", httpNodePort, replicaCount,
+          false, "/weblogic" + i);
+
+      int httpsNodePort = getApacheNodePort(domainNamespace, "https");
+      verifyClusterLoadbalancing(domainUids.get(i - 1), "", "https", httpsNodePort, replicaCount,
+          false, "/weblogic" + i);
+    }
+  }
+
+  private static void installIngressController() {
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java]: I changed the method name. However I did not move it to LoadBalancer Utility class. The methods contains Apache default usecase and custom usecase. It's specific to the test class.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+import static oracle.weblogic.kubernetes.utils.LoadBalancerUtils.installAndVerifyApache;
+import static oracle.weblogic.kubernetes.utils.OKDUtils.createRouteForOKD;
+import static oracle.weblogic.kubernetes.utils.OKDUtils.setTlsTerminationForRoute;
+import static oracle.weblogic.kubernetes.utils.OperatorUtils.installAndVerifyOperator;
+import static oracle.weblogic.kubernetes.utils.PersistentVolumeUtils.createPVPVCAndVerify;
+import static oracle.weblogic.kubernetes.utils.ThreadSafeLogger.getLogger;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+/**
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    createRouteForOKD("external-weblogic-operator-svc", opNamespace);
+    // Patch the route just created to set tls termination to passthrough
+    setTlsTerminationForRoute("external-weblogic-operator-svc", opNamespace);
+
+    // create pull secrets for WebLogic image when running in non Kind Kubernetes cluster
+    // this secret is used only for non-kind cluster
+    createSecretForBaseImages(domainNamespace);
+
+    for (int i = 1; i <= numberOfDomains; i++) {
+      domainUids.add("wls-domain" + i);
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java]: fixed.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+   * For details, please see
+   * https://github.com/oracle/weblogic-kubernetes-operator/tree/master/kubernetes/samples/charts/apache-samples/default-sample
+   */
+  @Test
+  @DisplayName("verify Apache load balancer default sample through HTTP channel")
+  void testApacheLoadBalancingDefaultSample() {
+    // verify Apache default sample
+    logger.info("Verifying Apache default sample");
+    int httpNodePort = getApacheNodePort(apacheNamespace, "http");
+    verifyClusterLoadbalancing(domainUids.get(0), "", "http", httpNodePort, replicaCount, false, "/weblogic");
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBApache.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+import static oracle.weblogic.kubernetes.utils.OKDUtils.setTlsTerminationForRoute;
+import static oracle.weblogic.kubernetes.utils.OperatorUtils.installAndVerifyOperator;
+import static oracle.weblogic.kubernetes.utils.SecretUtils.createSecretWithTLSCertKey;
+import static oracle.weblogic.kubernetes.utils.ThreadSafeLogger.getLogger;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+/**
+ * Test operator manages multiple domains.
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java (hunk 1 above)
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java (hunk 2 above)
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBNginx.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java (hunk 3 above)
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java (hunk 4 above)
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java (hunk 3 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java]: Test a single operator can manage multiple WebLogic domains with as single Traefik fronted loadbalancer --> 
Test a single operator can manage multiple WebLogic domains with a single Traefik fronted loadbalancer

**Review by anpanigr — COMMENTED** [review]
Can u add a - before the index  in domain id as follow ...

   domainUids.add("wls-nginx-domain" + i); to domainUids.add("wls-nginx-domain-" + i);
   domainUids.add("wls-apache-domain" + i); to domainUids.add("wls-apache-domain-" + i);
   domainUids.add("wls-traefik-domain" + i); to domainUids.add("wls-traefik-domain-" + i);

for better visibility during triaging

**Review by anpanigr — COMMENTED** [review]
Modify the script and pom files which refers to ItTwoDomainsLoadBalancers
I found one file oketest.sh

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java (hunk 3 above)
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItLBTraefik.java]: fixed

**Comment by xiancao** [comment]
> Can u add a - before the index in domain id as follow ...
> 
> domainUids.add("wls-nginx-domain" + i); to domainUids.add("wls-nginx-domain-" + i); domainUids.add("wls-apache-domain" + i); to domainUids.add("wls-apache-domain-" + i); domainUids.add("wls-traefik-domain" + i); to domainUids.add("wls-traefik-domain-" + i);
> 
> for better visibility during triaging

fixed

**Comment by xiancao** [comment]
> Modify the script and pom files which refers to ItTwoDomainsLoadBalancers
> I found one file oketest.sh

fixed

**Review by anpanigr — APPROVED** [review]
LGTM

**Review by maggiehe00 — APPROVED** [review]

**Pull request merged**